src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockDepths/instance_blackrock_depths.cpp

```diff
@@ -131,13 +138,30 @@ class instance_blackrock_depths : public InstanceMapScript
         uint32 TombTimer;
         uint32 TombEventCounter;
 
+        void UpdateMoira(Creature* moira)
+        {
+            InstanceMap::PlayerList const& players = instance->GetPlayers();
+
+            bool haveQuest = true;
+            for (Map::PlayerList::const_iterator i = players.begin(); i != players.end(); ++i)
+                if (Player * player = i->GetSource())
+                    if ((player->GetTeamId() == TEAM_ALLIANCE && !player->IsActiveQuest(QUEST_THE_PRINCESS_SURPRISE))
+                        || (player->GetTeamId() == TEAM_HORDE && !player->IsActiveQuest(QUEST_THE_PRINCESS_SAVED)))
+                        haveQuest = false;
```

Comment:
Wait a sec...
It will iterate all players of instance.
If last player in list, has no quest, the whole check method will skip...
Looks wrong to me.

Reply:
Yes, I think I forgot to break, if this is what you are telling.

Reply:
@Keader refactored the logic, see if you like this one more.